cloud/digitalocean/actuators/machine/userdata/ubuntu/templates.go

```diff
@@ -1,5 +1,267 @@
+    apt-get update
+
+    # If something failed during package installation but one of docker/kubeadm/kubelet was already installed
+    # an apt-mark hold after the install won't do it, which is why we test here if the binaries exist and if
+    # yes put them on hold
+    set +e
+    which docker && apt-mark hold docker docker-ce
+    which kubelet && apt-mark hold kubelet
+    which kubeadm && apt-mark hold kubeadm
+
+    # When docker is started from within the apt installation it fails with a
+    # 'no sockets found via socket activation: make sure the service was started by systemd'
+    # Apparently the package is broken in a way that it gets started without its dependencies, manually starting
+    # it works fine thought
+    which docker && systemctl start docker
+    set -e
+    
+    export CR_PKG=''
+{{- if .CRAptPackage }}
+{{- if ne .CRAptPackageVersion "" }}
+    export CR_PKG='{{ .CRAptPackage }}={{ .CRAptPackageVersion }}'
+{{- else }}
+    export CR_PKG='{{ .CRAptPackage }}'
+{{ end }}
+{{ end }}
+
+    # There is a dependency issue in the rpm repo for 1.8, if the cni package is not explicitly
+    # specified, installation of the kube packages fails
+    export CNI_PKG=''
+    {{- if semverCompare "=1.8.X" .KubernetesVersion }}
+    export CNI_PKG='kubernetes-cni=0.5.1-00'
+    {{- end }}
+    
+    DEBIAN_FRONTEND=noninteractive apt-get -o Dpkg::Options::="--force-confdef" -o Dpkg::Options::="--force-confold" install -y \
+      curl \
+      ca-certificates \
+      ceph-common \
+      cifs-utils \
+      conntrack \
+      e2fsprogs \
+      ebtables \
+      ethtool \
+      glusterfs-client \
+      iptables \
+      jq \
+      kmod \
+      openssh-client \
+      nfs-common \
+      socat \
+      util-linux \
+      ${CR_PKG} \
+      open-vm-tools \
+      kubelet={{ .KubernetesVersion }}-00 \
+      kubeadm={{ .KubernetesVersion }}-00 \
+      ${CNI_PKG}
+
+    PRIVATEIP=$(curl -s http://169.254.169.254/metadata/v1/interfaces/private/0/ipv4/address)
+    PUBLICIP=$(curl -s http://169.254.169.254/metadata/v1/interfaces/public/0/ipv4/address)
+
+    touch /tmp/kubernetes.kubeadm
+    cat << EOF  > "/tmp/kubernetes.kubeadm"
+    apiVersion: kubeadm.k8s.io/v1alpha1
+    kind: MasterConfiguration
+    token: {{ .BoostrapToken }}
+    kubernetesVersion: {{ .KubernetesVersion }}
+    nodeName: {{ .Machine.Name }}
+    api:
+      advertiseAddress: ${PUBLICIP}
+      bindPort: 443
+    apiServerCertSANs:
+    - ${PRIVATEIP}
+    - ${PUBLICIP}
+    - {{ .Machine.Name }}
+    authorizationModes:
+    - Node
+    - RBAC
+    EOF
+
+    cp /etc/default/kubelet-overwrite /etc/default/kubelet
+
+    systemctl enable --now docker
+    systemctl enable kubelet
+
+    kubeadm reset --force
+    kubeadm init --config /tmp/kubernetes.kubeadm
+
+    # Weave CNI plugin.
+    curl -SL "https://cloud.weave.works/k8s/net?k8s-version=$(kubectl version | base64 | tr -d '\n')&env.IPALLOC_RANGE=172.16.0.0/16" \
```

Comment:
Is there a solution for this upstream?

Having this in the cloud-init would force us to add support for every possible CNI provider which exists